include/bbp/sonata/compartment_sets.h

```diff
@@ -0,0 +1,209 @@
+#pragma once
+
+#include <bbp/sonata/nodes.h>
+
+#include <nlohmann/json.hpp>
+
+namespace bbp {
+namespace sonata {
+namespace detail {
+class CompartmentSetFilteredIterator;
+class CompartmentSet;
+class CompartmentSets;
+}  // namespace detail
+/**
+ * CompartmentLocation public API.
+ *
+ * This class uniquely identifies a compartment by a set of node_id, section_index and offset:
+ *
+ * - node_id: Global ID of the cell (Neuron) to which the compartment belongs. No
+ * overlaps among populations.
+ * - section_index: Absolute section index. Progressive index that uniquely identifies the section.
+ *  There is a mapping between neuron section names (i.e. dend[10]) and this index.
+ * - offset: Offset of the compartment along the section. The offset is a value between 0 and 1
+ */
+class SONATA_API CompartmentLocation
+{
+  private:
+    uint64_t node_id_ = 0;
+    uint64_t section_index_ = 0;
+    double offset_ = 0.0;
+
+    void setNodeId(int64_t node_id);
```

Comment:
When I suggested [POD](https://github.com/openbraininstitute/libsonata/pull/4/files/0bbaa3a7fac0ee7dfad6869e110be9afff2c16e7#r2128644704) I was generally thinking of a class without private members, getters, setters, etc to keep it at simple as possible - much like all the structs in config.h.
The checking of the invariants can happen when they are parsed and created.
Users don't need to be able to create them (or can create them with wrong data, but that's outside of the scope of libsonata.)

Reply:
> And... this was one of the initial solutions I proposed: do everything from the owner class.

I didn't understand that as such, sorry.

> However, this goes against encapsulation: even if we don’t use private members strictly, having a minimal interface avoids accidental misuse and clarifies intent.

Perhaps; IMO, the interface is the struct for something so simple.

> In some cases it can be easier to write unit tests for example.

In that scenario, I find it's easy enough to write the tests with the json inline; since the CompartmentSets can take a string, it's simple to create whatever `Location` you need.

How long would it take to try it? I think it's worth looking at to see how it behaves.

Reply:
> I didn't understand that as such, sorry.
No need to be sorry, I probably did not explain myself well (it was 1am)
> In that scenario, I find it's easy enough to write the tests with the json inline; since the CompartmentSets can take >a string, it's simple to create whatever Location you need.

>How long would it take to try it? I think it's worth looking at to see how it behaves.
I yiled. Probably 1h to be large. You do not need to create the full `CompartmentSets`. `CompartmentSet` already has a string constructor up and running